Fix sending the email with template 'bankwire' with parameters not set

LedCloud opened the pull request:
If one resend an email with template 'bankwire' or change the status of the order to that one, which is set to send an email with the template 'bankwire' the email is beeing send without setting the template variabled.

So the customer would receive email like that:
YOU HAVE SELECTED TO PAY BY WIRE TRANSFER.Here are the bank details for your transfer:
Amount: $76.51
Account owner: {bankwire_owner}
Account details: {bankwire_details}
Bank address: {bankwire_address}

More info could be read here http://forge.prestashop.com/browse/BOOM-4392

kpodemski commented:
@LedCloud it's a workaround but the problem is deeper, look at the ps_order_states table, i'm pretty sure that "module" table related to bankwire state is... `bankwire`, it should be `ps_wirepayment` tho

LedCloud commented:
> @LedCloud it's a workaround but the problem is deeper, look at the ps_order_states table, i'm pretty sure that "module" table related to bankwire state is... `bankwire`, it should be `ps_wirepayment` tho

Well. It's just a templape name. It could be anything else. Besides it might be disabled in the default state and binded to the other state, as it happended to me.

In my case I wanted to be sure that the stocks are really available and after that mannually set the other state to the order and send a email with a template "bankwire" correctly filled. And in this case those variables are not set. The same will happen if one wants to resend the state. (Push button "send again" in BO).

That's because this module sends a correct email only when the order is validated. when it is created.

Of course the variables name should be renamed to something that corrsponds to the module name, but it's not the problem, it's just a style I think.

My "workaround" is more flexible and will work in all cases. Sure it looks too simple :)

I think that the template should be moved to the module. Now it is inside a "system" folder /root/mails/... 
But it can not be moved to the module folder because you won't be able to select it in the settings of the desired state. Only templates from root/mails/ folder could be seleced there.

sarjon commented on ps_wirepayment.php:
```diff
@@ -92,7 +92,8 @@ public function __construct()
         Configuration::updateValue(self::FLAG_DISPLAY_PAYMENT_INVITE, true);
-        if (!parent::install() || !$this->registerHook('paymentReturn') || !$this->registerHook('paymentOptions')) {
+        if (!parent::install() || !$this->registerHook('paymentReturn') || !$this->registerHook('paymentOptions') ||
+        !$this->registerHook('actionGetExtraMailTemplateVars')) {
```
Don't you need to register new hook in `upgrade` file as well? :)

Btw, it would be nice if you could format this `if` into multiple lines for better readability.

LedCloud replied:
I'm not sure in myself )))) doing an upgrade needs a new version for this module. But sure you're right. There should be upgrade.

sarjon commented on ps_wirepayment.php:
```diff
@@ -442,4 +443,21 @@ public function getTemplateVarInfos()
         );
     }
+
+    public function hookActionGetExtraMailTemplateVars($params)
```
You can typehint with `array $params`.

sarjon commented on ps_wirepayment.php:
```diff
@@ -442,4 +443,21 @@ public function getTemplateVarInfos()
+
+    public function hookActionGetExtraMailTemplateVars($params)
+    {
+        if ($params['template']!='bankwire') {
```
Please add spaces, like `$params['template'] !== 'bankwire'`. :)

sarjon commented on ps_wirepayment.php:
```diff
@@ -442,4 +443,21 @@ public function getTemplateVarInfos()
+            return;
+        }
+
+        if (array_key_exists('{bankwire_owner}', $params['template_vars'])) {
```
Shouldn't you check on `extra_template_vars` instead of `template_vars`? :thinking:

LedCloud replied:
Nope. This hook is called to add some params and it's implied that everybody who wanted already set the variables and the vars set to $params['template_vars'].
Of course there could be intercection with other module, that can set somthing other to those variables, but what if that other module will be called later that this one?

sarjon commented on ps_wirepayment.php:
```diff
@@ -442,4 +443,21 @@ public function getTemplateVarInfos()
+            return;
+        }
+        
+        $params['extra_template_vars'] = array(
```
Is it safe to override everything that's inside `$params['extra_template_vars']`? There is a chance that other modules add data to it as well, right?

Btw, instead of defining new array of template vars, can you use `$this->extra_mail_vars`?

LedCloud replied:
Here is code from Mail.php
`
// Get extra template_vars
            $extraTemplateVars = [];
            Hook::exec(
                'actionGetExtraMailTemplateVars',
                [
                    'template' => $template,
                    'template_vars' => $templateVars,
                    'extra_template_vars' => &$extraTemplateVars,
                    'id_lang' => (int) $idLang,
                ],
                null,
                true
            );
`
I believe that I should change extra_template_vars or assign a new array to it.
Something like that:
'
$params['extra_template_vars']['{bankwire_owner}']  = Configuration::get('BANK_WIRE_OWNER');
//and so on
'
Both ways will work.

LedCloud commented:
By the way I think that those params can be included to PDF invoice as well. If someone, who wants to include them in PDF invoice, customize the PDF templates he count that {bankwire_owner} will be replaced by real values.

I will test that and if I succeed I'll do another PR and after that another one with a new version and upgrade with registering hook(s).

LedCloud commented:
I made an upgrade function, changed the version, changed the license headers in the files I edited.
Maybe someone should change the license in the other files too? I can do it in another PR.

sarahdib commented:
Hello @LedCloud 

We can't reproduce the problem the email is complete 
<img width="505" alt="Capture d’écran 2020-05-28 à 17 01 21" src="https://user-images.githubusercontent.com/3034677/83158510-27088680-a105-11ea-9620-4341c0d3ba13.png">

sarahdib commented:
@LedCloud 
I will close this pr, the problem seems fixed already.

If you reproduce the problem feel free to re-open it.

Have a nice day

ComonSoft commented:
Hi, the issue is still present when creating order in **backoffice**. The proposal fix with ActionGetExtraMailTemplateVars hook was perfect, why it is not present?
